File: 01-Arrays/best_time_to_buy_and_sell_stock.py
# """
# Problem Name: Best Time to Buy and Sell Stock
# Platform: LeetCode
# Difficulty: Easy
#
# Approach 1: Brute Force checking every pair. O(n^2) time.
# Approach 2: One-pass tracking the min price seen so far and calculating max profit.
# Optimal Approach: One-pass tracking min price.
#
# Time Complexity: O(n)
# Space Complexity: O(1)
#
# Pattern: Dynamic Programming / Greedy / Array
# """

from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
            
        min_price = float('inf')
        max_profit = 0
        
        for price in prices:
            if price < min_price:
                min_price = price
            else:
                profit = price - min_price
                max_profit = max(max_profit, profit)
                
        return max_profit

if __name__ == "__main__":
    sol = Solution()
    assert sol.maxProfit([7, 1, 5, 3, 6, 4]) == 5
    print("Best Time to Buy and Sell Stock Python: All tests passed! 🎉")

File: 01-Arrays/best_time_to_buy_and_sell_stock.cpp
/*
Problem Name: Best Time to Buy and Sell Stock
Platform: LeetCode
Difficulty: Easy

Approach 1: Brute Force calculating profit for all possible buy/sell day pairs (O(n^2) time).
Approach 2: Min-price tracker keeping track of the minimum price seen so far and maximizing profit.
Optimal Approach: One-pass Min-price tracker.

Time Complexity: O(n)
Space Complexity: O(1)

Pattern: Dynamic Programming / Greedy / Array
*/

#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
#include <cassert>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        if (prices.empty()) {
            return 0;
        }
        
        int min_price = INT_MAX;
        int max_profit = 0;
        
        for (int price : prices) {
            if (price < min_price) {
                min_price = price;
            } else {
                int profit = price - min_price;
                max_profit = std::max(max_profit, profit);
            }
        }
        
        return max_profit;
    }
};

int main() {
    Solution sol;
    std::vector<int> prices1 = {7, 1, 5, 3, 6, 4};
    assert(sol.maxProfit(prices1) == 5);
    std::cout << "Best Time to Buy and Sell Stock C++: All tests passed! 🎉" << std::endl;
    return 0;
}
